Added "The Good: Test" section to contribution guidelines (#1534)
kirbydesign/designsystem: .github/CONTRIBUTING.md

### Test

Tests helps us avoid regressions and provides peace of mind to other contributors when making changes to the code by ensuring that nothing is accidentally broken. Sometimes more time is spent on writing tests than the actual code - but it provides value everytime it is executed and is an important tool for collaboration.

> A well written test also acts as a sort of documentation by documenting functional requirements of the code.

Author: Changed it to: 
`A well written test also acts as documentation of functional requirements for the code.`

Using "of" twice gave it a weird rhythm :)

#### The good test is located in the right file

In Kirby we seperate integration tests from unit tests - so make sure to know which kind of testing you are doing before writing them. 

Unit tests are located in `*.spec.ts` files such as [ `button.component.spec.ts` ](https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/src/lib/components/button/button.component.spec.ts) and integration tests in `*.spec.integration.ts` files such as [ `button.component.integration.spec.ts` ](https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/src/lib/components/button/button.component.integration.spec.ts).

Remember that unit tests test a single unit - so if you find yourself relying on other components or functions without them being mocked or stubbed, then you are most likely writing an integration test. 

#### The good test can be read as a sentence 

Being able to read a test as a sentence, makes it clear what has gone wrong when the test fails and makes it act as a functional requirement. 

In the following two examples, note how example #2 clearly communicates intent while being easy to read and understand.

> Example #1 

Reviewer: @MadsBuchmann I meant adding it to the example code/image itself - currently the green checkmarks in the images hints at "good" examples:
![image](https://user-images.githubusercontent.com/9556693/120989595-ffb8b080-c77f-11eb-8c65-d9686fbe1982.png)

![](images/bad-test-example.png)

Example #2: 

![](images/good-test-example.png)

Example #1 can make sense when you have seen #2, but without you will have to interpret what is meant by "ButtonComponent + Kirby page Page Actions background-color: #fff".

> What is it exactly that should have the white backgound color here? 

Author: Reformatted the section: https://github.com/kirbydesign/designsystem/blob/documentation/1428-the-good-test/.github/CONTRIBUTING.md#the-good-test-can-be-read-as-a-sentence

> What is it exactly that should have the white backgound color here? 

Reviewer: Still a bit messy...

Not following this, causes the capitalisation to be wrong when the test is read as a sentence: 

> "ButtonComponent in Kirby Page Inside Page Actions Should render with correct background-color".

When done correctly it should read and be capitalized like a normal sentence:

> "ButtonComponent in Kirby Page inside Page Actions should render with correct background-color"

##### Avoid values in the description 

Notice how the above test says "should render with correct background-color" instead of "should render with background-color #fff". 

For someone running the tests, it is not relevant what the color should be, they care about if the color is _correct_. If they ever need to see what the correct color is, they can look up the test. 

This also reduces the likeliness of forgetting to correct the test description, if the color ever was to change - resulting in uncertainaity to what the correct color actually is. 

#### The good test follows the Arrange, Act, Assert (AAA) pattern

Following the [AAA pattern](https://medium.com/@pjbgf/title-testing-code-ocd-and-the-aaa-pattern-df453975ab80) makes your tests well structured and easy to understand. 

> 

Author: Done! https://github.com/kirbydesign/designsystem/blob/documentation/1428-the-good-test/.github/CONTRIBUTING.md#the-good-test-follows-the-arrange-act-assert-aaa-pattern

         'background-color': getColor('primary'),                                                                                                                                                                                                          
       });                                                                                                                                                                                                                                                 
  });
```

We do not denote each part of AAA with a comment but use spacing to seperate the different parts of the test.

You might notice tests in Kirby that do not use the AAA pattern - that is a good opportunity to do some girl/boy scouting and fix it if you are making other changes related to your issue in that file anyways.

#### The good test prioritizes readability above efficient code 

<!-- This section might be a bit redundant IMO -->

> When writing tests prioritise the readability of the test above all. 

Author: I'm not sure her royal highness queen Elizabeth II would agree - but it shall be done! 

![](https://i2-prod.liverpoolecho.co.uk/incoming/article17198667.ece/ALTERNATES/s615/1_JS187450175.jpg)

> When writing tests prioritise the readability of the test above all. 

Reviewer: Haha fair enough 😂 
But at least be consistent - it was written below the heading:
> The good test **prioritizes** readability above efficient code

It should be easy for future developers to visually parse and understand what is going on in the test. 

The following points can help you improve the readability of your tests:

* Make proper spacing between `it` blocks 
* Group related tests under common `describe` blocks 
* Keep your tests isolated (more on this here: [The good test is isolated and flat](#the-good-test-is-isolated-and-flat))
* Follow the AAA pattern as described above

Below there are two examples. Which one is easiest to read? 

![](images/unreadable-tests.png)

> ![](images/readable-tests.png)

Author: Reading it with fresh eyes it's actually a pretty poor example, you're right. 
I will find a better one!

> ![](images/readable-tests.png)

Author: Took in another example and added it as inline code instead :)

(Our guess is the second one)

#### The good test is isolated and flat

The tests should be isolated and flat, therefore writing tests is one of the few places, where it is okay to be [WET](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself#DRY_vs_WET_vs_AHA_solutions) instead of [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) - it is rather important actually. 

Future contributors should be able to read the test without having to jump around the code and do too much logic, therefore: 

* Avoid reusing code by using functions, keep your tests flat instead 
* Avoid the use of shared variables between tests, all state should be isolated to the test 
* Use `beforeEach` instead of `beforeAll`; it will cause potential changes made by previous tests to be overwritten, and you start with the same state each time

Following the above points makes it more likely that tests fail due to actual code changes and not how the tests are written. Changing a shared function can cause several tests to fail/change unexpectedly. If it happens too often that the test is the problem, not the code; it can cause mistrust in these.

> Therefore be WET - it will lower the chance of the tests being the problem and give a better overview of what is going on when reading the test. 

Author: Added a paragraph referring to test scenarios :) https://github.com/kirbydesign/designsystem/blob/documentation/1428-the-good-test/.github/CONTRIBUTING.md#the-good-test-is-isolated-and-flat

When doing unit tests, further isolation has to be done by stubbing and mocking everything else than what is being tested. 

#### The good test uses test scenarios when appropiate 

If you find yourself writing several tests that have identical arrange, act and assess sections - just with different variables, then you might benefit from using test scenarios. 

An array of scenarios are used to programatically create tests for each using the `forEach` function. Each test scenario contains the variables making up the scenario and then the expected outcome of this. 

Take for example the [ `kirby-button` ](https://cookbook.kirby.design/home/showcase/button) directive which is used to render buttons. Depending on which size is passed as an input property, it is rendered with different values for `font-size` , `height` and `min-width` . When testing this, the tests for each size is identical. They only differ in which value is given for size and the expected outcome. This can be expressed as an array of test scenarios: 

``` 

> const testScenarios: { size: ButtonSize; expected: any }[] = [

Reviewer: Consider either strong typing `expected` or removing the type altogether for simplicity of the example: 
```suggestion
const testScenarios = [
```

Author: Removed it for simplicities sake

    });

    it('should render with correct min-width', () => {
      expect(element).toHaveComputedStyle({ 'min-width': scenario.expected.minWidth });
    });
  });
});
```

This generates a total of 9 tests for us, but we only had to write 3! This is actually a simplified example taken from [ `button.component.spec.ts` ](https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/src/lib/components/button/button.component.spec.ts). In the actual file 9 tests are written; resulting in 27 tests being created programatically.

It is less WET than writing them all out by hand but each test is still flat, structured and self-contained. Actually it often reads better as the intention is clearer and gives a better overview of what is going on.

#### The good test prefers the use of Spectator over Angular testbed 

If you examine the test files, you will notice that almost every file uses the functions `createHostFactory` or `createComponentFactory` as part of their setup. These two functions are given the component being tested along with configuration such as declarations, imports, providers and more.

> They then respectively return a `createHost` or `createComponent` function that can be used to create a fresh component in your `beforeEach` blocks. 

Reviewer: ```suggestion
They then respectively return a factory function that can be used to create a fresh component in your `beforeEach` blocks. 
```

For example see: 

``` 

describe('ButtonComponent', () => {    
  let spectator: SpectatorHost<ButtonComponent>;    
  let element: HTMLButtonElement;    
    
  const createHost = createHostFactory({    
    component: ButtonComponent,    
    declarations: [MockComponent(IconComponent)],    
  });    
    
  describe('by default', () => {    
    beforeEach(() => {    
      spectator = createHost('<button kirby-button>Test</button>');    
      element = spectator.element as HTMLButtonElement;    
    });    
    
    it('should create', () => {    
      expect(spectator.component).toBeTruthy();    
    });
  })
})
```

Angular Test Bed is a nice tool for configuring and initializing the environment for unit tests. It unfortunately involves quite a bit of boilerplate code to use in testing. Therefore the use of Spectator is preferred.

> For more on `createHostFactory` , `createComponentFactory` , `createHost` and `createComponent` see [the Spectator documentation](https://github.com/ngneat/spectator#testing-components). 

Reviewer: ```suggestion
For more on `createHostFactory` and `createComponentFactory` see [the Spectator documentation](https://github.com/ngneat/spectator#testing-components). 
```

Remember that unit tests, tests a single unit - so if you find yourself relying on other components or functions without them being mocked or stubbed, then you are most likely writing an integration test. 

#### The good test can be read as a sentence 

Being able to read a test as a sentence, makes it clear what has gone wrong when the test fails and makes it act as a functional requirement. 

As an example, which one of these two tests best communicates intention while being easy to read and understand? 

Example #1: 

![](images/bad-test-example.png)

Example #2: 

![](images/good-test-example.png)

Probably #2. 

The first can make sense when you have seen #2, but without you will have to interpret what is actually meant by "ButtonComponent + Kirby page Page Actions background-color: #fff".

> What is it exactly that should have the white backgound color here? 

Author: I've made some changes in: d744f928539823e2078f55a8eb701e0bbec415c1

To achieve this, you can follow these pointers:

##### Use the [Given-When-Then](https://www.agilealliance.org/glossary/gwt/#q=~(infinite~false~filters~(postType~(~'page~'post~'aa_book~'aa_event_session~'aa_experience_report~'aa_glossary~'aa_research_paper~'aa_video)~tags~(~'given*20when*20then))~searchTerm~'~sort~false~sortDirection~'asc~page~1)) formula

For example have a look at example #1: 

* (Given) ButtonComponent in Kirby Page 
* (When) inside Page Actions 
* (Then) [it] should render with correct background-color

The missing _it_ comes from the `it()` blocks used when writing the tests: 
![](images/it-code-example.png)

##### Start `describe` and `it` blocks with a lowercase letter 

For example write `describe('inside Page Actions', () => {...})` not `describe('Inside Page Actions', () => {...})` .

Not following this, causes the capitalisation to be wrong when the test is read as a sentence: 

> > "ButtonComponent in Kirby Page Inside Page Actions Should render with correct background color".

Author: Hah good catch - will have to consider that one a bit :D

> > "ButtonComponent in Kirby Page Inside Page Actions Should render with correct background color".

Author: Add example displaying how it looks when done correctly

The following points can help you improve the readability of your tests:

* Make proper spacing between `it` blocks 
* Group related tests under common `describe` blocks 
* Keep your tests isolated (more on this here: [The good test is isolated](#the-good-test-is-isolated))
* Follow the AAA pattern as described above

Below there are two examples. Which one is easiest to read? 

![](images/unreadable-tests.png)

![](images/readable-tests.png)

(Our guess is the second one)

#### The good test is isolated and flat

The tests should be isolated and flat, therefore writing tests is one of the few places, where it is okay to be WET instead of DRY - it is rather important actually. 

Future contributors should be able to read the test without having to jump around the code and do too much logic, therefore: 

* Avoid reusing code by using functions, keep your tests flat instead 
* Avoid the use of shared variables between tests, all state should be isolated to the test 
* Use `beforeEach` instead of `beforeAll`; it will cause potential changes made by previous tests to be overwritten, and you start with the same state each time

> Not following the above points, makes it more likely that there is something wrong with the actual test and not the code being tested when they fail. Changing a shared function can cause several tests to fail/change unexpectedly. If it happens too often that the test is the problem, not the code; it can cause mistrust in these.

Author: Reworded: 1c9bcad317a177222b4d78f63dad9d367a63dac4

> Not following the above points, makes it more likely that there is something wrong with the actual test and not the code being tested when they fail. Changing a shared function can cause several tests to fail/change unexpectedly. If it happens too often that the test is the problem, not the code; it can cause mistrust in these.

Reviewer 2: Perfect

Therefore be WET - it will lower the chance of the tests being the problem and give a better overview of what is going on when reading the test. 

When doing unit tests, further isolation has to be done by stubbing and mocking everything else than what is being tested. 

#### The good test uses test scenarios when appropiate 

If you find yourself writing several tests that have identical arrange, act and assess sections - just with different variables, then you might benefit from using test scenarios. 

An array of scenarios are used to programatically create tests for each using the `forEach` function. Each test scenario contains the variables making up the scenario and then the expected outcome of this. 

> Take for example the [ `kirby-button` ](https://cookbook.kirby.design/home/showcase/button) directive which is used to render buttons. Depending on which size is passed as an input property, it is rendered with different values for `font-size` , `height` and `min-width` . When testing this, the tests for each size is identical. They only differ in which value is given for size and the expected outcome. This can be expressed as an array of test scenarios: 

Reviewer 2: Just saw that you *do* link to the test file below 😊

> Take for example the [ `kirby-button` ](https://cookbook.kirby.design/home/showcase/button) directive which is used to render buttons. Depending on which size is passed as an input property, it is rendered with different values for `font-size` , `height` and `min-width` . When testing this, the tests for each size is identical. They only differ in which value is given for size and the expected outcome. This can be expressed as an array of test scenarios: 

Author: Is it okay if I just leave it as is - I purposely mention the actual file afterwards as to not overwhelm the reader, and instead show a simplified example 😄

#### The good test prefers fakeAsync over Async & Done 

While the `it` function is provided with a `done` callback as an argument and support the `async` / `await` syntax for testing asynchronous behavior; we encourage the use of the [ `fakeAsync` ](https://angular.io/api/core/testing/fakeAsync) function instead. 

Using the `done` function might send you straight to [callback hell](https://image.slidesharecdn.com/promisesandchaininginangularjs-141027044455-conversion-gate02/95/promises-and-chaining-in-angularjs-into-callback-hell-and-back-again-17-638.jpg) while the `async` / `await` syntax slows down execution of tests. 

`fakeAsync` combined with the [ `tick` ](https://angular.io/api/core/testing/tick) function on the other hand, simulates the asynchronous passage of time without actually taking any additional time and avoiding callback hell.

When using `fakeAsync` remember to not mix it with the `done` function or the `async` / `await` syntax. Using `fakeAsync` we want to simulate that time is passing - not actually wait for the time to pass. 

#### The good test uses ionicModuleForTest instead of IonicModule

> When working with ionic you might have to import the `IonicModule` as part of your `createComponent` or `createHost` factory.   

Reviewer 2: "ionic" should probably be "Ionic" with a capital `I`

Here you should use the [ `TestHelper` ](https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/src/lib/testing/test-helper.ts) property `ionicModuleForTest` like so: 

``` 

const createComponent = createComponentFactory({
  component: RadioComponent,
  imports: [TestHelper.ionicModuleForTest],
});
```

This ensures that each test uses the same config for the `IonicModule` . 

There might be files where `IonicModule` is used directly instead of `TestHelper.ionicModuleForTest` - this is a good chance to do some girl/boy scouting and fix it, but only if you are making changes to those files anyways.  

#### The good test combines tests when appropiate

> In Kirby there is defined a [ `toHaveComputedStyle` ](https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/src/lib/testing/element-css-custom-matchers.d.ts) custom matcher that checks the value of CSS properties an element is rendered with, like so: 

Reviewer 2: Minor suggestion for change:

> In Kirby there is a toHaveComputedStyle custom matcher defined that checks the value of CSS properties an element is rendered with [...]

or 

> Kirby has a toHaveComputedStyle custom matcher that checks the value of CSS properties an element is rendered with [...]

Author: Went with the first one 😄

``` 

it('should render with correct border', () => {
  expect(element).toHaveComputedStyle({
    'border-width': '1px',
    'border-style': 'solid',
    'border-color': 'transparent'
  });
});
```

This makes the intention more clear and also saves some time writing tests. Should the test fail due to one of the properties being rendered with an unexpected value; the error message will display which property is wrong. 

We are aware that this requires a bit of gut feel. Testing for example if an element has `position: relative` & `background-color: #ffffff` would most likely not make sense. What would the test say? _Should have correct position and background-color_? There is really no relation between these two properties.

Testing the properties `position` , `left` , `right` , `bottom` and `top` could make better sense as it could be tested as _"should be positioned correctly"_. As an additional example `font-family` , `font-size` , `font-weight` & `font-style` could be tested as _"should have correct typography"_.

A rule of thumb is to test properties together whenever they have the same prefix such as `font-` . 

> This will however not always be the case as can be seen in the positioning example.

Reviewer 2: Having a single rule of thumb example immediately followed by an exception makes the rule of thumb a bit weak 😊

Would it make sense to be more specific e.g. by mentioning that you probably often want to test individual properties of [shorthand properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Shorthand_properties#see_also) together?

For example `border` is shorthand for `border-width`, `border-style` and `border-color` so you would create a *border* test (just like the example above).

Similarly, `font` is also a shorthand property, which calls for at *font* test that combines assessment of individual font-* properties.

In addition to mentioning shorthand properties you could mention *related properties*, like the position/top/right/bottom/left example. Other examples of related properties are:

* **Color**: `color` and `background-color` which often *should* be tested together
* **Typography**: `font-*`, `line-height`, `letter-spacing` and `text-*` properties
* **Positioning**: `position`, `top/right/bottom/left` and `z-index`

Author: Let's have a chat about this one 😄

### The good test is located in the right file

In Kirby we seperate integration tests from unit tests - so make sure to know which kind of testing you are doing before writing them. 

Unit tests are located in `*.spec.ts` files such as [ `button.component.spec.ts` ](https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/src/lib/components/button/button.component.spec.ts) and integration tests in `*.spec.integration.ts` files such as [ `button.component.integration.spec.ts` ](https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/src/lib/components/button/button.component.integration.spec.ts).

Remember that unit tests test a single unit - so if you find yourself relying on other components or functions without them being mocked or stubbed, then you are most likely writing an integration test. 

### The good test can be read as a sentence 

Being able to read a test as a sentence, makes it clear what has gone wrong when the test fails and makes it act as a functional requirement. 

Have a look at the following two examples:

```Javascript

> describe('ButtonComponent + Kirby Page', () => {

Reviewer: ```suggestion
// Example #1
describe('ButtonComponent + Kirby Page', () => {
```

  describe('Page Actions', () => {
    it('background-color: #fff', ...) 
  })
})
```

```Javascript

> describe('ButtonComponent in Kirby Page', () => {

Reviewer: ```suggestion
// Example #2
describe('ButtonComponent in Kirby Page', () => {
```

  describe('inside page actions', () => {
    it('should render with correct background-color', ...) 
  })
})
```

> Start by reading the first example as a sentence: _"ButtonComponent + Kirby Page Page Actions background-color: #fff"_. Notice how hard the test is to read and understand when read like this. The intent of the test is also unclear; what is it exactly that should have the white background color here? Page Actions? ButtonComponent? Kirby Page?

Reviewer: Consider rendering the sentence as a quote as is done in the `Start describe and it blocks with a lowercase letter` section below

> Now read the second example as a sentence: _"ButtonComponent in Kirby Page inside page actions should render with correct background-color"_. It can easily be read as a sentence and the intent of the test is clear. There is no guessing what should have the white background color applied; the sentence clearly states that it is the ButtonComponent. 

Reviewer: Consider rendering the sentence as a quote as is done in the `Start describe and it blocks with a lowercase letter` section below

Being able to read a test as a sentence, makes it clear what has gone wrong when the test fails and makes it act as a functional requirement. 

In the following two examples - note how the _good example_ clearly communicates intent while being easy to read and understand.

> Consider the following _bad_ example: 

Reviewer: This is a bit messy...
Maybe something along the lines of: "Consider the following two examples. In the first example... In the second example notice how..."

Author: Took another stab at it. Not exactly what you proposed, but something along the lines of that 😄

```

```Javascript
// Example #2
describe('ButtonComponent in Kirby Page', () => {

>   describe('inside page actions', () => {

Reviewer: Align casing with example code in following section:
```suggestion
  describe('inside Page Actions', () => {
```

Following the [AAA pattern](https://medium.com/@pjbgf/title-testing-code-ocd-and-the-aaa-pattern-df453975ab80) makes your tests well structured and easy to understand. 

An example of this pattern can be found in the file [`list-helper.spec.ts`](https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/src/lib/components/list/helpers/list-helper.spec.ts): 

```javascript
it('should emit load more event, if load on demand is enabled and is not loading', () => {
  // Arrange 
  component.isLoadOnDemandEnabled = true;
  component.isLoading = false;    
                                                        
  // Act 
  listHelper.onLoadOnDemand(component, null);    

  // Assert
  expect(component.loadOnDemand.emit).toHaveBeenCalledTimes(1);
});
```

> We do not necessarily denote each part of AAA with a comment as in the above example (it is okay if you do) but at least use spacing to seperate the different parts of the test. 

Reviewer: Typo:
```suggestion
We do not necessarily denote each part of AAA with a comment as in the above example (it is okay if you do) but at least use spacing to separate the different parts of the test. 
```

* Keep your tests isolated (more on this here: [The good test is isolated and flat](#the-good-test-is-isolated-and-flat))
* Follow the AAA pattern as described above

Below are two examples. Which one is easiest to read and understand? 

> The following where: 

Reviewer: Not sure I understand this sentence/structure? 🤔

Author: Took a stab at restructuring it a bit: https://github.com/kirbydesign/designsystem/pull/1534/commits/c2ab9b6e11af37116c3ba220b3c51721ec094013

    expect(component.loadOnDemand.emit).not.toHaveBeenCalled();
  });
});
```

Or this one where: 
* There are spacing between `it` blocks
* Related tests are grouped under common describes 
* Tests are isolated due to the use of `beforeEach`
* The AAA pattern is followed

```javascript 
describe('ListHelper function: OnLoadOnDemand', () => {
  let listHelper: ListHelper; 
  let component: ListComponent; 

  beforeEach(() => {

> 

Reviewer: A little to much line spacing :)
```suggestion
```

    listHelper = new ListHelper(); 
    component = {
      loadOnDemand: new EventEmitter<LoadOnDemandEvent>(),
    } as ListComponent;
    component.loadOnDemand.subscribe((loadMoreEvent: LoadOnDemandEvent) => {});

> 

Reviewer: A little to much line spacing :)
```suggestion
```

  }); 

  describe('when load on demand is disabled', () => {

> 

Reviewer: A little to much line spacing :)
```suggestion
```

        expect(component.loadOnDemand.emit).not.toHaveBeenCalled();
      });
    });

> 

Reviewer: A little to much line spacing :)
```suggestion
```

```javascript 
describe('ButtonComponent', () => {    
  let spectator: SpectatorHost<ButtonComponent>; 
  let element: HTMLButtonElement; 

> 

Reviewer: A little to much line spacing :)
```suggestion
```

    

  const createHost = createHostFactory({    

    component: ButtonComponent,    
    declarations: [MockComponent(IconComponent)],    

> 

Reviewer: A little to much line spacing :)
```suggestion
```

  }); 

    

  describe('by default', () => {    

> 

Reviewer: A little to much line spacing :)
```suggestion
```

    beforeEach(() => {    
      spectator = createHost('<button kirby-button>Test</button>'); 
      element = spectator.element as HTMLButtonElement; 
    }); 

    

> 

Reviewer: A little to much line spacing :)
```suggestion
```

    it('should create', () => {    
      expect(spectator.component).toBeTruthy(); 
    }); 

  })
})

> 

Reviewer: A little to much line spacing :)
```suggestion
```
